File: hs/17.hs
letters n | n < 20 = ones !! n
          | mod n 10 == 0 && n < 100 = tens !! (quot n 10)
          | n < 100 = tens !! (quot n 10) + ones !! (mod n 10)
          | True = ones !! (quot n 100) + (if mod n 100 == 0 then 7 else 10) + letters (mod n 100)
    where ones = [0,3,3,5,4,4,3,5,5,4,3,6,6,8,8,7,7,9,8,8]
          tens = [0,3,6,6,5,5,5,7,6,6]

main = print $ (foldl1 (+) $ map letters [1..1000]) + 1

File: clj/17.clj
(def ones [0 3 3 5 4 4 3 5 5 4 3 6 6 8 8 7 7 9 8 8])
(def tens [0 3 6 6 5 5 5 7 6 6])

(defn letters [n]
  (cond
    (< n 20) (nth ones n)
    (and (= 0 (mod n 10)) (< n 100)) (nth tens (/ n 10))
    (< n 100) (+ (nth tens (/ n 10))
                 (nth ones (mod n 10)))
    :else (+ (nth ones (/ n 100))
             (if (= 0 (mod n 100)) 7 10)
             (letters (mod n 100)))))

(prn (inc (apply + (map letters (range 1 1001)))))
